File: hello-swift/Sources/Client/Core/ChunkManager.swift
import Foundation

/// Chunk manager for handling file chunking operations
class ChunkManager {
    static let defaultChunkSize = 8192 // 8KB to avoid WebSocket frame fragmentation
    
    /// Calculate the number of chunks needed for a file
    static func calculateChunkCount(fileSize: Int64, chunkSize: Int = defaultChunkSize) -> Int {
        return Int((fileSize + Int64(chunkSize) - 1) / Int64(chunkSize))
    }
    
    /// Get the size of a specific chunk
    static func getChunkSize(fileSize: Int64, offset: Int64, defaultSize: Int = defaultChunkSize) -> Int {
        return min(defaultSize, Int(fileSize - offset))
    }
    
    /// Validate chunk parameters
    static func validateChunk(fileSize: Int64, offset: Int64, chunkSize: Int) throws {
        guard offset >= 0 && offset < fileSize else {
            throw NSError(domain: "ChunkManager", code: 1, 
                         userInfo: [NSLocalizedDescriptionKey: "Invalid offset: \(offset)"])
        }
        guard chunkSize > 0 else {
            throw NSError(domain: "ChunkManager", code: 2, 
                         userInfo: [NSLocalizedDescriptionKey: "Invalid chunk size: \(chunkSize)"])
        }
    }
}

File: hello-rust/src/client/chunk_manager.rs
// Chunk manager for handling file chunking operations

#[allow(dead_code)]
pub const CHUNK_SIZE: usize = 65536; // 64KB

#[allow(dead_code)]
pub struct ChunkManager;

#[allow(dead_code)]
impl ChunkManager {
    pub fn calculate_chunk_count(file_size: u64, chunk_size: usize) -> usize {
        ((file_size + chunk_size as u64 - 1) / chunk_size as u64) as usize
    }

    pub fn get_chunk_size(file_size: u64, offset: u64, default_chunk_size: usize) -> usize {
        std::cmp::min(default_chunk_size as u64, file_size - offset) as usize
    }
}
